### PR TITLE
[SPARK-36646][SQL] Push down group by partition column for aggregate

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
lift the restriction for aggregate push down for parquet and orc if group by columns are the same as the partition cols

### Why are the changes needed?
previously, if there are group by columns, we don't push down aggregate to data source.
After the change, if the group by columns are the same as the partition columns, we will push down aggregates.

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
new tests

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala
Comment (c21) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:461]: nit: `partition col values` -> `partition columns values`
nit: `result set row ` -> `result row`? What's `result set row`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala
Comment (c21) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:501]: nit: `orcValuesDeserializer.deserializeFromValues(aggORCValues)` can be called only once before the if-else branch.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala
Comment (c21) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala:273]: shall we also add test for aggregate on group-by column? e.g. `count(p)`, `max(p)` etc.

**Review by c21 — COMMENTED** [review]
Thanks @huaxingao. I think core logic looks good, just have some minor comments. cc @viirya.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala
Comment (huaxingao) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala]: I added `count(p)`, but didn't add `max(p)` since it won't be pushed down

**Comment by huaxingao** [comment]
@c21 Thanks a lot for review! I have addressed the comments. Please take one more look when you have time. Thanks!
also cc @sunchao @viirya

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/AggregatePushDownUtils.scala
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/AggregatePushDownUtils.scala:151]: nit: `groupByColumns`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:462]: nit: `schemaWithoutGroupBy`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala:205]: nit: maybe use `aggregation.groupByColumns.nonEmpty`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala
Comment (sunchao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala:264]: nit: add JIRA number

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/AggregatePushDownUtils.scala
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/AggregatePushDownUtils.scala:148]: nit: maybe `getSchemaWithoutGroupColumns`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala:206]: just curious how we make sure the `partitionValues` is always consistent with `aggregation.groupByColumns`? in terms of ordering.

**Review by sunchao — COMMENTED** [review]
looks good just a few nits

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/AggregatePushDownUtils.scala
Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/AggregatePushDownUtils.scala:151]: changed to `numOfGroupByColumns`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala
Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:462]: changed

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala
Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala:205]: changed

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala
Comment (huaxingao) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala:264]: I want it to be consistent with the rest of the tests, so didn't add a jira number

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala
Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala:206]: I actually forgot to considering ordering. Fixed. Thanks for pointing this out!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/AggregatePushDownUtils.scala
Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/AggregatePushDownUtils.scala:148]: Changed to `getSchemaWithoutGroupingExpression` per this suggestion https://github.com/apache/spark/pull/34445#discussion_r740418455

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetPartitionReaderFactory.scala
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetPartitionReaderFactory.scala:138]: can we call this in `ParquetUtils.createAggInternalRowFromFooter`? it seems we always call `AggregatePushDownUtils.reOrderPartitionCol` then followed by `ParquetUtils.createAggInternalRowFromFooter`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/AggregatePushDownUtils.scala
Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/AggregatePushDownUtils.scala:175]: Seems to me that the `partitionValues` comes from `PartitionPath`, which always contains `GenericInternalRow`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/AggregatePushDownUtils.scala
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/AggregatePushDownUtils.scala:154]: nit: maybe swap the order of `aggSchema` and `aggregation` here, as we're modifying the schema here with the info from aggregation.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala
Comment (sunchao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala:312]: nit: remove this?

**Review by sunchao — COMMENTED** [review]
Thanks @huaxingao , looks good, just a few more nits after which it's ready to merge I think.

**Review by sunchao — APPROVED** [review]
LGTM pending CI, thanks @huaxingao ! it'd be great if you can add a bit more details in the PR description too.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala:271]: Hmm, can you test both `aggPushDownEnabledKey` as true and false and see if the results are the same?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala:303]: here too. We should make sure `aggPushDownEnabledKey` won't change results.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcPartitionReaderFactory.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcPartitionReaderFactory.scala:90]: Seems `filePath` can be created after the `if` block:

```scala
if (aggregation.nonEmpty) {
  return buildReaderWithAggregates(file, conf)
}

val filePath = new Path(new URI(file.filePath))
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcPartitionReaderFactory.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcPartitionReaderFactory.scala:134]: ditto.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala:264]: Don't need check case sensitivity now?

Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala:264]: seems to me no need to check case sensitivity because I have normalized aggregates and group by columns in `V2ScanRelationPushDown`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceAggregatePushDownSuite.scala:306]: Hmm, if we enable aggregate push down later one day, this test might be ignorantly missed to its original goal. Should we explicitly set `aggPushDownEnabledKey` as false and collect the expected results?

**Review by viirya — COMMENTED** [review]
Looks good. One remaining question about test.

**Review by c21 — APPROVED** [review]
LGTM

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Last commit is test-only and GA was passed. Merging to master. Thanks.

**Comment by huaxingao** [comment]
Thank you all

**Pull request closed without merging**
